### ISSUE TITLE
Can't get user token when using external auth

### ISSUE THREAD

**Issue by SteelAlloy** [Chocobozzz/PeerTube#6258]
### Describe the current behavior

To retrieve a token that can be used via the api, you need to go through the endpoint `api/v1/users/token` according to the [documentation](https://docs.joinpeertube.org/api/rest-getting-started#get-user-token).

However, when you change the authentication mode to something other than local auth, e.g. `peertube-plugin-auth-openid-connect`, this no longer works.

And as this endpoint requires a password, I don't see how to proceed

### Steps to reproduce

1. Create a user, set password manually or via the API
2. Change auth plugin for this user
3. Try to get a user API token

### Describe the expected behavior

Two solutions :
- The password method works when a password is set via the API
- Another endpoint to create user token for external auth plugins

### Additional information

* PeerTube instance:
  * URL:
  * Version: 6.0.2
  * NodeJS version:
  * Ffmpeg version:

* Browser name, version and platforms on which you could reproduce the bug:
* Link to browser console log if relevant:
* Link to server log if relevant (`journalctl` or `/var/www/peertube/storage/logs/`):

**Comment by Chocobozzz** [comment]
Can you describe the command/HTTP call that doesn't work? `/api/v1/oauth-clients/local` is not related to user and do not require a password

**Comment by Chocobozzz** [comment]
Users with external auth plugin that uses classic `username`/`password` (like LDAP) can use the classic `api/v1/users/token` endpoint.

But remote external auth plugins (like `openid-connect` or `saml`) requires the user to authenticate on their own web page. After a successful authentication, the remote auth platform redirects the user on your PeerTube instance login page with a `externalAuthToken` query param, used by the PeerTube client that sends it to the server.

Can you detail your use case?

**Comment by SteelAlloy** [comment]
I'm writing a script for my university to facilitate mass video upload.
Those videos are uploaded to one channel only, so i could switch the auth plugin for this user each time I run the script, but this is not ideal

**Comment by kbrackney** [comment]
Any movement on this feature enhancement.  The OpenID PeerTube plugin works great in conjunction with the PeerTube client login process.   However, I am also looking for a way to directly return valid access and refresh tokens from PeerTube api for an external auth user that does not rely on the redirection to the client path `/login?externalAuthToken=${bypassToken}&username=${user.username}`. Is there a way to authenticate an external user that solely interacts with the API?

**Comment by Chocobozzz** [comment]
No sorry, we need to implement https://tools.ietf.org/html/rfc8628

**Comment by jakmeier** [comment]
Hello, I am also interested in this feature.

My use case:

- I'd like to use PeerTube as a video management backend. My website / PWA embeds PeerTube videos and should now also let users upload videos directly, without switching to PeerTube. Videos are recorded directly inside my application and ideally I want to upload them directly from the browser's memory to PeerTube.
- For user management, I use Keycloak. Using [peertube-plugin-auth-openid-connect](https://framagit.org/framasoft/peertube/official-plugins/tree/master/peertube-plugin-auth-openid-connect) to create PeerTube accounts for my existing users worked well for me. However, those users cannot use any of the authenticated API endpoints on PeerTube due to the limitation described in this issue.

Question 1: Is this use case (access from an external website / PWA) generally in scope of the project? If you want, I might be able to try and contribute the necessary changes if it's just a matter of developer time. But if PeerTube is not really intended to be integrated like that, I might be better off looking for other video hosting software.

Question 2: Does this really need the device flow (https://tools.ietf.org/html/rfc8628)? I think this is for OAuth 2.0 clients without a browser.

I would have thought more of something like a token exchange (https://tools.ietf.org/html/rfc8693) that lets my app retrieve a PeerTube OAuth token for an externally authenticated user.

Or maybe even allow the use of external tokens on API points and do a token introspection there. (http://tools.ietf.org/html/rfc7662) I imagine token introspection is already done on the PeerTube backend when it gets the forwarded `externalAuthToken` from the PeerTube client, so this might be easy to add.

But obviously, I'm not familiar with the architecture of PeerTube and its plugin system and don't know what makes the most sense. I'm happy with anything that gets the job done securely.

**Comment by Chocobozzz** [comment]
> Is this use case (access from an external website / PWA) generally in scope of the project?

Yes it is!

> Does this really need the device flow (https://tools.ietf.org/html/rfc8628)? I think this is for OAuth 2.0 clients without a browser.

Maybe not, clients without a browser are quite rare. We can first focus to provide tokens to CLI (with a web browser available), mobile app' and external websites.

In the current architecture:
 1. The frontend sends the user to a dedicated route, managed by the openid plugin
 2. The openid plugin redirects the user to the external provider
 3. Ask the external provider to redirect the user in a plugin route
 4. The plugin checks the provider response and call a peertube core function `userAuthenticated`
 5. This function redirects the user to `peertube.example.com/login?specialAuthToken=...` web page
 6. The PeerTube frontend uses the special `specialAuthToken` and sends it back to the REST API to generate internal PeerTube access/refresh tokens

What if the frontend in 1. can ask PeerTube to redirect on a special URL, instead or `peertube.example.com/login?specialAuthToken=...`?  Of course we would have to use a redirect URL allow list, that admins can update.

With this method, mobile apps can use their scheme to redirect users (`myapp://...`) and your web app can redirect users on your web app URL.

**Comment by jakmeier** [comment]
@Chocobozzz  Thanks for the response!

> What if the frontend in 1. can ask PeerTube to redirect on a special URL, instead or peertube.example.com/login?specialAuthToken=...? Of course we would have to use a redirect URL allow list, that admins can update.
> 
> With this method, mobile apps can use their scheme to redirect users (myapp://...) and your web app can redirect users on your web app URL.

Let me see if I understand that right. To get an API access token:

1. My web app redirects to the openid plugin url in PeerTube
2. The plugin redirects to the external identity provider (Keycloak)
3. The external provider produces an OpenID connect token (`externalToken`) and redirects back to PeerTube
4. The plugin verifies `externalToken`, calls `userAuthenticated`, and creates a `specialAuthToken`
5. Now the plugin redirects to my web app with the `specialAuthToken`
6. My web app uses the `specialAuthToken` to retrieve an access token from PeerTube (`PeertubeAccessToken`) through an existing REST API endpoint
7. Now the web app can make API calls using `PeertubeAccessToken`

Is that what you suggest?

Could we also implement an API route in the openid connect plugin that directly takes an `externalToken` and returns a `specialAuthToken` that way?  Maybe that could avoid the redirects in steps 1 and 3.

**Comment by Chocobozzz** [comment]
> Is that what you suggest?

Yes it is!

> Could we also implement an API route in the openid connect plugin that directly takes an externalToken and returns a specialAuthToken that way? Maybe that could avoid the redirects in steps 1 and 3.

I'm not sure because the openid plugin must generate the verifier codes to check the keycloak answer (but my knowledge of openid/oauth is limited, so I may be wrong).

**Comment by jakmeier** [comment]
> I'm not sure because the openid plugin must generate the verifier codes to check the keycloak answer (but my knowledge of openid/oauth is limited, so I may be wrong).

Yes, if PeerTube is going through the (PKCE) Authorization Code Flow, then you are right. And maybe that is a good idea to implement it that way, since the oidc plugin already can do this.

The thing is, my users already go through this process using my web app backend for the code exchange. So they already have an access token from that. Ideally, for the full SSO experience, users wouldn't need also log in to Keycloak through PeerTube.

I believe in theory they could just use the existing token to authenticate PeerTube, if the PeerTube plugin can verify the validity of the existing token and link it to the corresponding internal user. In OAuth terms, the PeerTube plugin would be a passive resource server that only validates tokens.

Okta has [this article](https://www.oauth.com/oauth2-servers/token-introspection-endpoint/) where they describe how an OAuth 2.0 client (my web app) can make a call to a resource server (PeerTube) with a token generated by the authorization server (Keycloak). No need to go through any authorization flows again.

Using [RFC 7662](https://datatracker.ietf.org/doc/html/rfc7662), a PeerTube plugin could make a network call to Keylocak to verify the token, without generating a new one. Or with [RFC-9068](https://datatracker.ietf.org/doc/html/rfc9068), there is not even a need for a network call for each request, only the signer public key of Keycloak needs to be fetched. Both are already implemented on Keycloak. The plugin just needs to implement the client side on the new API endpoint that I suggested.

Does that make sense?

**Comment by Chocobozzz** [comment]
Yes it does, thanks for the detailed explanations! I suggest we resolve the OP issue first, using the workflow described in https://github.com/Chocobozzz/PeerTube/issues/6258#issuecomment-3136693741.

And then contribute to the [openid peertube plugin][https://framagit.org/framasoft/peertube/official-plugins/-/tree/master/peertube-plugin-auth-openid-connect?ref_type=heads) to add a dedicated plugin route that would be able to resolve your use case.

**Comment by jakmeier** [comment]
Makes sense! I will try work to work on it over the next couple of weeks as much as I can find the time for it. Can't promise how much that is, though.

**Comment by jakmeier** [comment]
@Chocobozzz I have a working local setup with the flow you described, where I was able to make an API call to get my own user info for a user registered over Keycloak OpenID connect.

The server side changes are in #7179. It''s still a draft because I haven't looked into writing a test, yet.

The plugin code still lacks configuration of the allowed urls but and might also need tests.

To submit the plugin changes, I registered on framagit.org and I am awaiting approval. I'm assuming I would need to send a merge request there to contribute to the plugin, right?

edit: plugin draft merge request is now also online: https://framagit.org/framasoft/peertube/official-plugins/-/merge_requests/32

**Comment by Chocobozzz** [comment]
https://framagit.org/framasoft/peertube/official-plugins/-/merge_requests/32 has been merged and will be compatible with PeerTube 7.3 (thanks to https://github.com/Chocobozzz/PeerTube/pull/7179).

Thanks again @jakmeier
